src/main/java/bio/terra/cloudfunctions/common/CloudStorageEventHarness.java

```diff
@@ -0,0 +1,67 @@
+package bio.terra.cloudfunctions.common;
+
+import com.google.cloud.functions.BackgroundFunction;
+import com.google.cloud.functions.Context;
+import com.google.gson.internal.LinkedTreeMap;
+import java.util.logging.Logger;
+
+/**
+ * This class is an abstract representation of a Function triggered by a Google Cloud Storage Event.
+ * The function casts the Google Cloud Storage Event into a Map of key-value pairs.
+ *
+ * <p>This class contains only framework-specific logic to receive and parse Google Storage Object
+ * Event messages.
+ *
+ * <p>Sub-classes of this class can be integrated with a DI framework to deploy Cloud Functions like
+ * a Service.
+ */
+public abstract class CloudStorageEventHarness implements BackgroundFunction<LinkedTreeMap<?, ?>> {
+  private static final Logger logger = Logger.getLogger(CloudStorageEventHarness.class.getName());
+
+  private Context context;
+  private LinkedTreeMap<?, ?> event;
+
+  /**
+   * @param event String
+   * @param context event function context
+   * @throws Exception when something goes wrong
+   */
+  @Override
+  public void accept(LinkedTreeMap<?, ?> event, Context context) throws Exception {
+    this.event = event;
+    this.context = context;
+    doAccept();
+  }
+
```

Comment:
I think this class should implement `RawBackgroundFunction` and store the event as a string and then deserialize from that string in `getEvent` What do we gain by using the LinkedTreeMap to hold the event data?  Here's a suggestion for the class: 
```suggestion
public abstract class CloudStorageEventHarness implements RawBackgroundFunction {
  private static final Logger logger = Logger.getLogger(CloudStorageEventHarness.class.getName());

  private Context context;
  private String event;


  @Override
  public void accept(String event, Context context) throws Exception {
    this.event = event;
    this.context = context;
    doAccept();
  }


  /** @return Google Cloud Storage Event Context */
  public Context getContext() {
    return context;
  }

  /**
   * @param classOfT generic type of event pojo
   * @param <T> generic event pojo
   * @return an event pojo of generic type T matching the provider's specification of the event. *
   *     (e.g. GCSEvent is a pojo of GCS StorageObjectData event). Other providers have equivalent
   *     events * such as AWS S3Event and Azure BlobTrigger. Different providers use slightly
   *     different naming and * syntax for the event trigger (e.g. accept for GCS, handleRequest for
   *     S3, run for Azure Blob * Storage)
   */
  public <T> T getEvent(Class<T> classOfT) {
    try {
      return GsonWrapper.convertFromClass(this.event, classOfT);
    } catch (JsonSyntaxException e) {
      logger.log(Level.SEVERE, "Could not convert from JSON", e);
      throw e;
    }
  }

  /**
   * Subclasses of this class must implement this method to kick off application-specific functions
   * downstream .
   *
   * @throws Exception when something wrong happened
   */
  public abstract void doAccept() throws Exception;
}
```

Reply:
I think we need a final vote on which base class we would like to use as it seems we are going back and forth between `BackgroundFunction<T>` and `RawBackgroundFunction`.

Let me describe these base classes below then we can have a vote on which cloud function representation we feel more comfortable to use.

- With `BackgroundFunction<T>`, an unbounded generic type `T` is not allowed. 
- Under the hood, BackgroundFunction<T> tries to cast the event json to generic type `T`, which is not possible due to runtime type erasure. Instead, `BackgroundFunction<Map<K,V>>` use `Map<K,V>` as the fundamental representation of an event. Cast `Map<K,V>` to any event pojo such as `GcsEvent` using the `getEvent(Class)` method.
- With `RawBackgroundFunction`, the event json is represented as a pure Java String which we can parse using our `GsonWrapper`. 
- Compare `BackgroundFunction<Map<K,V>>` with `RawBackgroundFunction`. The `BackgroundFunction<Map<K,V>>` has a more efficient representation. Both `BackgroundFunction<Map<K,V>>` and `RawBackgroundFunction` can use our `GsonWrapper` with overloaded `convertFromClass` methods.